File: Day 04/01-Block的回顾/01-Block的回顾/HttpTools.m
//
//  HttpTools.m
//  01-Block的回顾
//
//  Created by 1 on 16/9/27.
//  Copyright © 2016年 小码哥. All rights reserved.
//

#import "HttpTools.h"


@interface HttpTools()

@property(nonatomic ,copy) void (^finishedCallback)(NSString *);

@end

@implementation HttpTools

- (void)loadData:(void (^)(NSString *))finishedCallback
{
    self.finishedCallback = finishedCallback;
    
    // 1.发送网络的异步请求
    dispatch_async(dispatch_get_global_queue(0, 0), ^{
        NSLog(@"已经发送了网络请求:%@", [NSThread currentThread]);
        
        // 2.回到主线程
        dispatch_sync(dispatch_get_main_queue(), ^{
            NSLog(@"回到主线程:%@", [NSThread currentThread]);
            
            // 3.获取数据, 并且将数据回调出去
            finishedCallback(@"json数据");
        });
        
    });
    
}

@end

File: Day 04/02-闭包的使用/02-闭包的使用/HttpTools.swift
//
//  HttpTools.swift
//  02-闭包的使用
//
//  Created by 1 on 16/9/27.
//  Copyright © 2016年 小码哥. All rights reserved.
//

import UIKit

class HttpTools {
    
    var finishedCallback : ((_ jsonData : String) -> ())?
    
    // 闭包类型: (_ 参数列表) -> (返回值类型)
    // @escaping : 逃逸的
    func loadData(_ finishedCallback : @escaping (_ jsonData : String) -> ()) {
        
        self.finishedCallback = finishedCallback
        
        // 1.发送异步网络请求
        DispatchQueue.global().async {
            print("发送异步网络请求:\(Thread.current)")
            
            DispatchQueue.main.sync {
                print("回到主线程:\(Thread.current)")
                
                finishedCallback("json数据")
            }
        }
    }
}
